mint-client/src/clients/user.rs

```diff
@@ -420,3 +452,190 @@ impl From<LnClientError> for ClientError {
+
+        while let Some(key) = access.next_key::<String>()? {
+            match key.as_str() {
+                "address" => addr = Some(access.next_value()?),
+                "amount" => amt = Some(access.next_value()?),
+                _ => (),
+            }
+        }
+        Ok(PegOutReq {
+            address: bitcoin::Address::from_str(addr.unwrap()).unwrap(), //unwraps are not ok here I think but I will come back to Errors in another commit anyway
+            amount: bitcoin::Amount::from_sat(amt.unwrap()),
+        })
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for WrappedInvoice {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_str(InvoiceVisitor {})
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for PegOutReq {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_map(PegOutReqVisitor {})
+    }
+}
+///Holds either a Success or an Error variant and will be send as a response to the client (note: this means all [`ClientError`] will be send as [`tide::Result`] Ok
+/// only other errors will cause the server to respond with [`tide::Result`] Err)
+pub enum Event {
+    Success(APIResponse),
+    Error(APIResponse), //TODO: define Error APIResponse variant, implement a build method and use it in the endpoints
+}
+/// Holds all possible Responses of the RPC-CLient can also be used to parse responses (for client-cli)
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub enum APIResponse {
+    ///The clients holdings : The quantity of coins for each tier. For total holdings sum(Infoi.quantity * Infoi.tier) with i = 0 - n
+    /// Also contains the [`APIResponse::Pending`] variant.
+    Info {
+        coins: Vec<CoinsByTier>,
+        pending: Box<APIResponse>,
+    },
+    /// Active issuances : Not yet (bey the federation) signed BUT accepted coins
+    Pending {
+        //TODO: Also return Vec<TransactionId> (?)
+        transactions: usize,
+        acc_qty_coins: usize,
+        acc_val_amount: Amount,
+    },
+    ///Holds a new address for the client to use for peg-in
+    PegInAddress { pegin_address: String },
+    ///Holds a [`minimint_api::TransactionId`] from a successful PegIn or PegOut
+    PegIO { txid: String },
+    /// Holds the serialized [`Coins<SpendableCoin>`]
+    Spend { token: Coins<SpendableCoin> },
+    /// Holds the from the federation returned [`OutPoint`] (regarding the reissuance) and the [`TransactionStatus`]
+    Reissue {
+        out_point: OutPoint,
+        status: TransactionStatus,
+    },
+    /// Holds events which could not be sent to the client but were triggered by some action from him. This will be cleared after querying it
+    EventDump { events: Vec<APIResponse> },
+    /// Represents an event which occurred. Might be an Error or Non-Error
+    Event { time: u64, msg: String },
```

Comment:
Maybe we can have typed events? As an enum? Would be interesting what you envision as events.

Reply:
I just wanted to make sure there is a way to store errors which occurred after someone already got some response. I addet a "successful fetch" event too but that is maybe too verbose since the absence of an error implicates a success. This is very costly since the deamon would have all the events in memory and they grow until they get dumped (/events). Also it uses Mutex which is costly too... Maybe the clientd should just log such events and don't make them accessible over API ?